clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/PubSubBasedAdminClientManager.java

```diff
@@ -114,74 +103,80 @@ private Future<SubscriptionAdminClient> getOrCreateSubscriptionAdminClient() {
         try {
             final Topic createdTopic = client.createTopic(topic);
             if (createdTopic == null) {
                 LOG.debug("Creating topic failed [topic: {}, projectId: {}]", topic, projectId);
                 return Future.failedFuture("Topic creation failed.");
             }
-            topics.add(createdTopic.getName());
-            return Future.succeededFuture(topic.getTopic());
-        } catch (AlreadyExistsException ex) {
+            LOG.debug("Topic {} created successfully.", topic);
             return Future.succeededFuture(topic.getTopic());
         } catch (ApiException e) {
             LOG.debug("Error creating topic {} on project {}", topic, projectId, e);
             return Future.failedFuture("Topic creation failed.");
         }
-
     }
 
     /**
-     * Gets an existing subscription or creates a new one on Pub/Sub based on the given subscription endpoint and prefix.
+     * Gets an existing subscription or creates a new one on Pub/Sub based on the given subscription endpoint and
+     * prefix.
      *
      * @param endpoint The endpoint name of the subscription, e.g. command_internal.
      * @param prefix The prefix of the subscription, e.g. the adapter instance ID.
-     * @return A succeeded Future if the subscription is successfully created or already exists, or a failed
-     *         Future if it could not be created.
+     * @return A succeeded Future if the subscription is successfully created or already exists, or a failed Future if
+     *         it could not be created.
      */
     public Future<String> getOrCreateSubscription(final String endpoint, final String prefix) {
         final String topicAndSubscriptionName = PubSubMessageHelper.getTopicName(endpoint, prefix);
         final TopicName topic = TopicName.of(projectId, topicAndSubscriptionName);
         final SubscriptionName subscription = SubscriptionName.of(projectId, topicAndSubscriptionName);
 
-        if (subscriptions.contains(subscription.toString())) {
-            LOG.debug("Subscription {} already exists, continue", subscription);
-            return Future.succeededFuture(subscription.getSubscription());
-        }
         return getOrCreateSubscriptionAdminClient()
                 .onFailure(thr -> LOG.debug("admin client creation failed", thr))
-                .compose(client -> getOrCreateSubscription(projectId, subscription, topic, client));
+                .compose(client -> getSubscription(subscription, client)
+                        .recover(thr -> createSubscription(projectId, subscription, topic, client)));
+    }
+
+    private Future<String> getSubscription(final SubscriptionName subscription, final SubscriptionAdminClient client) {
+        try {
+            return Future.succeededFuture(client.getSubscription(subscription).getName());
+        } catch (ApiException e) {
+            return Future.failedFuture("Could not get topic");
```

Comment:
we usually include the exception that caused the error so that we can log a stack trace if needed ...

Reply:
Makes sense. I added it.